cmd/migrate/main.go

```diff
@@ -0,0 +1,161 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	conversions "github.com/kubermatic/machine-controller/pkg/api/conversions"
+	downstreammachineclientset "github.com/kubermatic/machine-controller/pkg/client/clientset/versioned"
+	"github.com/kubermatic/machine-controller/pkg/controller"
+	downstreammachines "github.com/kubermatic/machine-controller/pkg/machines"
+
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	clusterv1alpha1clientset "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+
+	apiextclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/retry"
+)
+
+var (
+	kubeconfig string
+	masterURL  string
+)
+
+func main() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flag.Parse()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	if err != nil {
+		glog.Fatalf("Error building kubeconfig: %v", err)
+	}
+
+	apiExtClient := apiextclient.NewForConfigOrDie(cfg)
+	clusterv1alpha1Client := clusterv1alpha1clientset.NewForConfigOrDie(cfg)
+	kubeClient := kubernetes.NewForConfigOrDie(cfg)
+
+	if err = migrateIfNecesary(kubeClient, apiExtClient, clusterv1alpha1Client, cfg); err != nil {
```

Comment:
how do we want to design that tool ? can we tolerate partial errors ? for example what if we successfully migrated 9 out of 10 machines ?

Reply:
Then it is supposed to fail and try again. If it can not migrate the `machine-controller` will not come up as this runs in an `initContainer`. If that case every occurs an operator has to check what the issue is.